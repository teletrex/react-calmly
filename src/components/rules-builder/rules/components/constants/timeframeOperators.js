/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../../constants';

const { NONE, RANGE, SINGLE_VALUE, WHEN_RELATION, WITHIN_PERIOD } = RULE_INPUT_TYPES;

export const TIMEFRAME_OPERATORS = [
  // Within:
  {
    inputType: WITHIN_PERIOD,
    label: 'within the last',
    value: 'WITHIN_THE_LAST',
  },

  // Today:
  {
    inputType: NONE,
    label: 'on today',
    value: 'ON_TODAY',
  },
  {
    inputType: NONE,
    label: 'before today',
    value: 'BEFORE_TODAY',
  },

  // Any time:
  {
    inputType: NONE,
    label: 'at any time',
    value: 'AT_ANY_TIME',
  },

  // Date:
  {
    inputType: SINGLE_VALUE,
    label: 'on',
    value: 'ON',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'after',
    value: 'AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'before',
    value: 'BEFORE',
  },

  // Date range:
  {
    inputType: RANGE,
    label: 'between',
    value: 'BETWEEN',
  },

  // When relation:
  {
    inputType: WHEN_RELATION,
    label: 'exactly',
    value: 'EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'more than',
    value: 'MORE_THAN',
  },
  {
    inputType: WHEN_RELATION,
    label: 'less than',
    value: 'LESS_THAN',
  },
];
