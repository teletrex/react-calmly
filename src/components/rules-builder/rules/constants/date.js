/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DATE_PATTERN = 'd{1,2}/d{1,2}/d{4}';
