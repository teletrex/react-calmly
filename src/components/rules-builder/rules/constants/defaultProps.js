/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DEFAULT_PROPS = {
  INPUT_FIELD_VALIDATION: [],
  OPERATORS: [],
  TIMEFRAME: {},
};
