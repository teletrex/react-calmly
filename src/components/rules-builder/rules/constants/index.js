/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './date';
export * from './defaultProps';
export * from './periods';
export * from './propTypes';
export * from './ruleOperators';
export * from './validationMessages';
export * from './values';
export * from './whenRelations';
