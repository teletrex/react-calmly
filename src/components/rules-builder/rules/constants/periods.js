/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const PERIODS = [
  {
    label: 'days',
    value: 'DAYS',
  },
  {
    label: 'weeks',
    value: 'WEEKS',
  },
  {
    label: 'months',
    value: 'MONTHS',
  },
];
