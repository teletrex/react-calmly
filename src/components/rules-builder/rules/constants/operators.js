/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../constants';

const {
  MULTIPLE_VALUES,
  NONE,
  PERIOD_NUMBER,
  PERIOD_RANGE,
  RANGE,
  SINGLE_VALUE,
  WHEN_RELATION,
  WITHIN_PERIOD,
} = RULE_INPUT_TYPES;

// Common operators
export const BLANK_OPERATORS = [
  {
    inputType: NONE,
    label: 'is blank',
    value: 'IS_BLANK',
  },
  {
    inputType: NONE,
    label: 'is not blank',
    value: 'IS_NOT_BLANK',
  },
];

export const EXACT_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'is exactly',
    value: 'IS_EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'is not exactly',
    value: 'IS_NOT_EXACTLY',
  },
];

export const IS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is',
    value: 'IS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is not',
    value: 'IS_NOT',
  },
];

export const HAS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'has',
    value: 'HAS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'has not',
    value: 'HAS_NOT',
  },
];

export const IS_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'is one of the following',
    value: 'IS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'is not one of the following',
    value: 'IS_NOT_ONE_OF_THE_FOLLOWING',
  },
];

export const CONTAIN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'contains',
    value: 'CONTAINS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not contain',
    value: 'DOES_NOT_CONTAIN',
  },
];

export const CONTAIN_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'contains one of the following',
    value: 'CONTAINS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not contain one of the following',
    value: 'DOES_NOT_CONTAIN_ONE_OF_THE_FOLLOWING',
  },
];

export const RANGE_OPERATORS = [
  {
    inputType: RANGE,
    label: 'is between',
    value: 'IS_BETWEEN',
  },
];

// String related operators
export const STARTS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'starts with',
    value: 'STARTS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not start with',
    value: 'DOES_NOT_START_WITH',
  },
];

export const STARTS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'starts with one of the following',
    value: 'STARTS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not start with one of the following',
    value: 'DOES_NOT_START_WITH_ONE_OF_THE_FOLLOWING',
  },
];

export const ENDS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'ends with',
    value: 'ENDS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not end with',
    value: 'DOES_NOT_END_WITH',
  },
];

export const ENDS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'ends with one of the following',
    value: 'ENDS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not end with one of the following',
    value: 'DOES_NOT_END_WITH_ONE_OF_THE_FOLLOWING',
  },
];

// Number related operators
export const LESS_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is less than',
    value: 'IS_LESS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is less than or equal to',
    value: 'IS_LESS_OR_EQUAL',
  },
];

export const GREATER_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than',
    value: 'IS_GREATER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than or equal to',
    value: 'IS_GREATER_OR_EQUAL',
  },
];

// Date related operators
export const TODAY_OPERATORS = [
  {
    inputType: NONE,
    label: 'is today',
    value: 'IS_TODAY',
  },
  {
    inputType: NONE,
    label: 'is after today',
    value: 'IS_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not after today',
    value: 'IS_NOT_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is before today',
    value: 'IS_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not before today',
    value: 'IS_NOT_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'anniversary is today',
    value: 'ANNIVERSARY_IS_TODAY',
  },
];

export const ANNIVERSARY_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'anniversary is',
    value: 'ANNIVERSARY_IS',
  },
];

export const WITHIN_OPERATORS = [
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the last',
    value: 'IS_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the last',
    value: 'IS_NOT_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the next',
    value: 'IS_WITHIN_THE_NEXT',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the next',
    value: 'IS_NOT_WITHIN_THE_NEXT',
  },
];

export const WITHIN_RANGE_OPERATORS = [
  {
    inputType: PERIOD_RANGE,
    label: 'is within the range of the last',
    value: 'IS_WITHIN_THE_RANGE_OF_THE_LAST',
  },
];

export const PERIOD_EQUAL_OPERATORS = [
  {
    inputType: PERIOD_NUMBER,
    label: 'year is',
    value: 'YEAR_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'month is',
    value: 'MONTH_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'day is',
    value: 'DAY_IS',
  },
];

export const BEFORE_AFTER_DATE_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is after',
    value: 'IS_AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is before',
    value: 'IS_BEFORE',
  }
]

export const DATA_AGGREGATION_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'by',
    value: 'BY',
  },

];
