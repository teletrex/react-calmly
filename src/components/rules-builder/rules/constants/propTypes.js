/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import PropTypes from 'prop-types';

import { RULE_INPUT_TYPES } from '../../constants';

const { any, arrayOf, func, oneOf, oneOfType, shape, string } = PropTypes;

const OPERATOR = oneOfType([
  string,
  shape({
    inputType: oneOf(Object.values(RULE_INPUT_TYPES)).isRequired,
    label: string.isRequired,
    value: string.isRequired,
  }),
]);
const OPERATORS = arrayOf(OPERATOR);

const TIMEFRAME = shape({
  data: shape({
    operator: string,
    value: any,
  }),
  onChange: func,
});

const OPERATOR_NAME = string;

const SELECT_OPTION = shape({
  label: string.isRequired,
  value: string.isRequired,
});

const SELECT_OPTIONS = arrayOf(SELECT_OPTION);

export const PROPTYPES = {
  OPERATOR,
  OPERATORS,
  OPERATOR_NAME,
  SELECT_OPTION,
  SELECT_OPTIONS,
  TIMEFRAME,
};
