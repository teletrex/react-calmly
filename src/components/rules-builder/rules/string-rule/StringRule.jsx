/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import PropTypes from 'prop-types';

import { RULE_INPUT_TYPES } from '../../constants';
import { DEFAULT_PROPS, PROPTYPES, VALUES } from '../constants';
import { getRuleConfigData, getRuleId } from '../helpers';

import { OneOfValues, TextValue } from './inputFields';

const StringRule = ({ config, id, onValueChange, operator, value }) => {
  const { placeholder } = getRuleConfigData(config);

  const commonProps = {
    id,
    onValueChange,
    operator,
    value,
  };

  switch (operator?.inputType) {
    case RULE_INPUT_TYPES.SINGLE_VALUE:
      return <TextValue {...commonProps} placeholder={placeholder.textInput} />;

    case RULE_INPUT_TYPES.MULTIPLE_VALUES:
      return <OneOfValues {...commonProps} placeholder={placeholder.addTagsInline} />;

    default:
      return null;
  }
};

StringRule.propTypes = {
  config: PropTypes.shape({
    operators: PROPTYPES.OPERATORS,
    placeholder: PropTypes.shape({
      addTagsInline: PropTypes.string,
      textInput: PropTypes.string,
    }),
  }),
  id: PropTypes.string,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
};

StringRule.defaultProps = {
  config: {
    operators: DEFAULT_PROPS.OPERATORS,
    placeholder: {},
  },
  id: getRuleId('string'),
  operator: null,
  value: VALUES.UNPROVIDED,
};

export default StringRule;
