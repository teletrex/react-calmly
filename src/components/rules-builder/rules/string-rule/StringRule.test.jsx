/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render } from '@testing-library/react';

import { renderWithContext } from '../../utils/testHelpers';
import { IS_OPERATORS, IS_ONE_OF_OPERATORS } from '../constants/operators';

import StringRule from './StringRule';

const defaultInitialOperator = IS_OPERATORS[0];
const defaultInitialValue = 'abc';

describe('StringRule', () => {
  const onValueChange = jest.fn();
  let defaultProps;

  beforeEach(() => {
    onValueChange.mockReset();

    defaultProps = {
      id: 'TestStringRuleElement',
      onValueChange,
      operator: defaultInitialOperator,
      value: defaultInitialValue,
    };
  });

  it('should render correctly', () => {
    const { container } = render(<StringRule {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container } = renderWithContext(<StringRule {...defaultProps} />, { isReadOnly: true });
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only and oneOfValues operator', () => {
    const props = {
      ...defaultProps,
      operator: IS_ONE_OF_OPERATORS[0],
      value: ['Foo', 'Bar'],
    };
    const { container } = renderWithContext(<StringRule {...props} />, { isReadOnly: true });
    expect(container).toMatchSnapshot();
  });

  it('should properly render the one of values component', () => {
    const initialValue = ['test', 'abc'];
    const props = {
      ...defaultProps,
      operator: IS_ONE_OF_OPERATORS[0],
      value: initialValue,
    };
    const { getByPlaceholderText, getByText } = render(<StringRule {...props} />);

    expect(getByText(initialValue[0])).toBeInTheDocument();
    expect(getByText(initialValue[1])).toBeInTheDocument();
    expect(getByPlaceholderText('Enter value')).toBeInTheDocument();
  });

  it('should not render any values if the provided operator is empty', () => {
    const props = {
      ...defaultProps,
      operator: null,
    };
    const { container } = render(<StringRule {...props} />);

    expect(container.innerHTML).toBe('');
  });
});
