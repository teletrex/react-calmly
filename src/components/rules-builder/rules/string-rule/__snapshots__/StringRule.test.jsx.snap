// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringRule should render correctly 1`] = `
<div>
  <div
    class="bx--form-item bx--text-input-wrapper"
  >
    <span
      class="bx--text-input__label-wrapper bx--visually-hidden"
    >
      <label
        class="bx--label"
        for="TestStringRuleElement"
      >
        String rule
      </label>
      
    </span>
    <div
      class="bx--text-input__field-wrapper"
    >
      <input
        class="bx--text-input"
        id="TestStringRuleElement"
        name="TestStringRuleElement"
        placeholder="Enter value"
        type="text"
        value="abc"
      />
    </div>
  </div>
</div>
`;

exports[`StringRule should render correctly for read-only 1`] = `
<div>
  <span
    class="bx--rule-value-readonly bx--text-space-left"
  >
    abc
  </span>
</div>
`;

exports[`StringRule should render correctly for read-only and oneOfValues operator 1`] = `
<div>
  <span
    class="bx--rule-value-readonly bx--text-space-left"
  >
    Foo, Bar
  </span>
</div>
`;
